services/test_analytics/tests/test_ta_process_flakes.py

```diff
@@ -0,0 +1,280 @@
+        uploads=[
+            {
+                "state": "processed",
+                "testruns": [
+                    {"test_id": "test1", "outcome": "pass"},
+                    {"test_id": "test2", "outcome": "pass"},
+                    {"test_id": "test3", "outcome": "failure"},
+                    {"test_id": "test4", "outcome": "flaky_failure"},
+                    {"test_id": "test5", "outcome": "error"},
+                    {"test_id": "test6", "outcome": "skip"},
+                ],
+            }
+        ],
+        existing_flakes=[
+            {"test_id": "test1", "count": 5, "fail_count": 2},
+        ],
+    )
+
+    process_flakes_for_repo(result["repoid"])
+
+    flakes = {
+        bytes(flake.test_id).decode(): {
+            "count": flake.count,
+            "fail_count": flake.fail_count,
+            "recent_passes_count": flake.recent_passes_count,
+        }
+        for flake in Flake.objects.all()
+    }
+
+    assert snapshot("json") == flakes
+
+
+def test_update_existing_flakes(setup_test_data):
+    result = setup_test_data(
+        uploads=[
+            {
+                "state": "processed",
+                "testruns": [
+                    {"test_id": "test1", "outcome": "pass"},
+                    {"test_id": "test1", "outcome": "failure"},
+                ],
+            }
+        ],
+        existing_flakes=[
+            {
+                "test_id": "test1",
+                "count": 5,
+                "fail_count": 2,
+                "recent_passes_count": 0,
+            },
+        ],
+    )
+
+    process_flakes_for_repo(result["repoid"])
+
+    flake = Flake.objects.get(test_id=b"test1")
+    assert flake.count == 7
+    assert flake.fail_count == 3
+    assert flake.recent_passes_count == 0
+
+
+def test_create_new_flakes(setup_test_data):
+    result = setup_test_data(
+        uploads=[
+            {
+                "state": "processed",
+                "testruns": [
+                    {"test_id": "test1", "outcome": "failure"},
+                    {"test_id": "test2", "outcome": "flaky_failure"},
+                    {"test_id": "test3", "outcome": "error"},
+                ],
+            }
+        ],
+        existing_flakes=[],
+    )
+
+    process_flakes_for_repo(result["repoid"])
+
+    assert Flake.objects.count() == 3
+    for test_id in [b"test1", b"test2", b"test3"]:
+        flake = Flake.objects.get(test_id=test_id)
+        assert flake.count == 1
+        assert flake.fail_count == 1
+        assert flake.recent_passes_count == 0
+        assert flake.start_date is not None
+        assert flake.end_date is None
+
+
+def test_flake_expiry_and_recreation(setup_test_data):
+    result = setup_test_data(
+        uploads=[
+            {
+                "state": "processed",
+                "testruns": [
+                    {"test_id": "test1", "outcome": "pass"},
+                    {"test_id": "test1", "outcome": "failure"},
```

Comment:
testing with 2x `pass` would trigger the error I mentioned above, as you clear the test from the current flakes, and on the second iteration try to access that yet again.